Spigot-Server-Patches/0697-Introduce-beacon-activation-deactivation-events.patch

```diff
@@ -0,0 +1,52 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Spyridon Pagkalos <spyridon@ender.gr>
+Date: Thu, 25 Mar 2021 20:28:04 +0200
+Subject: [PATCH] Introduce beacon activation/deactivation events
+
+
+diff --git a/src/main/java/net/minecraft/world/level/block/entity/TileEntityBeacon.java b/src/main/java/net/minecraft/world/level/block/entity/TileEntityBeacon.java
+index 8dfb9eb12d07c2d4737ecf3de1ae7f6de31891bf..8e8ab3d3148cd152dd0d9de5bbd02d34e7d5a0c3 100644
+--- a/src/main/java/net/minecraft/world/level/block/entity/TileEntityBeacon.java
++++ b/src/main/java/net/minecraft/world/level/block/entity/TileEntityBeacon.java
+@@ -46,6 +46,8 @@ import org.bukkit.potion.PotionEffect;
+ import org.bukkit.craftbukkit.event.CraftEventFactory;
+ import org.bukkit.entity.Player;
+ import com.destroystokyo.paper.event.block.BeaconEffectEvent;
++import io.papermc.paper.event.block.BeaconActivatedEvent;
++import io.papermc.paper.event.block.BeaconDeactivatedEvent;
+ // Paper end
+ 
+ public class TileEntityBeacon extends TileEntity implements ITileInventory, ITickable {
+@@ -211,6 +213,10 @@ public class TileEntityBeacon extends TileEntity implements ITileInventory, ITic
+                 boolean flag1 = this.levels > 0;
+ 
+                 if (!flag && flag1) {
++                    // Paper start - BeaconActivatedEvent
++                    org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(world, position);
++                    new BeaconActivatedEvent(block, (org.bukkit.block.Beacon) block.getState()).callEvent();
++                    // Paper end
+                     this.a(SoundEffects.BLOCK_BEACON_ACTIVATE);
+                     Iterator iterator = this.world.a(EntityPlayer.class, (new AxisAlignedBB((double) i, (double) j, (double) k, (double) i, (double) (j - 4), (double) k)).grow(10.0D, 5.0D, 10.0D)).iterator();
+ 
+@@ -220,6 +226,10 @@ public class TileEntityBeacon extends TileEntity implements ITileInventory, ITic
+                         CriterionTriggers.l.a(entityplayer, this);
+                     }
+                 } else if (flag && !flag1) {
++                    // Paper start - BeaconDeactivatedEvent
++                    org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(world, position);
++                    new BeaconDeactivatedEvent(block, (org.bukkit.block.Beacon) block.getState()).callEvent();
++                    // Paper end
+                     this.a(SoundEffects.BLOCK_BEACON_DEACTIVATE);
+                 }
+             }
+@@ -257,6 +267,10 @@ public class TileEntityBeacon extends TileEntity implements ITileInventory, ITic
+ 
+     @Override
+     public void al_() {
++        // Paper start - BeaconDeactivatedEvent
++        org.bukkit.block.Block block = org.bukkit.craftbukkit.block.CraftBlock.at(world, position);
```

Comment:
This method is called when a tile entity is removed, so the block here will no longer be a beacon.

Reply:
Hello!
First of all, nice catch! 😅 
I have two strategies to solve this. I can either:
* Never fire the event when breaking the beacon, making people use the break event for these cases, or
* Create a `CraftBeacon` myself using the `TileEntityBeacon` and return that in the event.
Which one do you thing I should go for?

Reply:
I'd personally just not provide a beacon in the event; just the block reference and tell people to check themselves if it's (still) a beacon. Not firing the event when breaking sounds counterintuitive and the second option would be a bit of a hack

Reply:
I'll go with the nullable route, sounds like a better strategy!

Reply:
Fix pushed, thanks for the review again!